llvm_passes/HipPrintf.cpp

```diff
@@ -0,0 +1,203 @@
+// LLVM IR pass to convert calls to the CUDA/HIP printf() to OpenCL/SPIR-V
+// compatible ones.
+//
+// (c) 2021 Pekka Jääskeläinen / Parmance for Argonne National Laboratory
+//
+// SPIRV-LLVM translator generates a wrong pointer address space to printf
+// format string if it's not the correct (constant) one in the input. This
+// pass moves the format string to constant address space before we pass the IR
+// to SPIRV emission.
+
+#include "HipPrintf.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+#include <iostream>
+
+using namespace llvm;
+
+class HipPrintfToOpenCLPrintfLegacyPass : public ModulePass {
+public:
+  static char ID;
+  HipPrintfToOpenCLPrintfLegacyPass() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override {
+    return false;
+  }
+
+  StringRef getPassName() const override {
+    return "Convert printf calls to OpenCL compatible ones.";
+  }
+};
+
+char HipPrintfToOpenCLPrintfLegacyPass::ID = 0;
+static RegisterPass<HipPrintfToOpenCLPrintfLegacyPass>
+    X("hip-printf", "Convert printf calls to OpenCL compatible ones.");
+
+// Converts the address space of the format string to OpenCL compatible one
+// by creating a copy of it in the module global scope.
+//
+// Also counts the number of format args for replacing the return
+// value of the printf() call with it for CUDA emulation.
+Value* convertFormatString(Value *HipFmtStrArg, Instruction *Before,
+                           unsigned &NumberOfFormatSpecs) {
+
+  Module *M = Before->getParent()->getParent()->getParent();
+
+  Type *Int8Ty = IntegerType::get(M->getContext(), 8);
+  ConstantExpr *CE = cast<ConstantExpr>(HipFmtStrArg);
+  assert(CE->isGEPWithNoNotionalOverIndexing());
+
+  // TODO: Cannot we really get access to the ptr operand in a simpler way
+  // without a new object to delete? Perhaps use the cexpr operand replace API?
+  GetElementPtrInst *GEP =
+    cast<GetElementPtrInst>(CE->getAsInstruction());
+  GlobalVariable *OrigFmtStr =
+    cast<GlobalVariable>(GEP->getPointerOperand());
+
+  Constant *FmtStrData = OrigFmtStr->getInitializer();
+
+  NumberOfFormatSpecs = 0;
+  int I = 0;
+  while (Constant *Chr = FmtStrData->getAggregateElement(I)) {
```

Comment:
If the `FmtStrData` is a ConstantDataSequential you could use its getAsString() and count `%`s and `%%`s with StringRef::count().

Reply:
Should work better as it indeed is. I tried to look for such a method but overlooked, it seems. Thanks!